File: built-in plugins dsp/mb-util-oscilloscope.cpp
/**
 * MB Oscilloscope
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Real-time waveform display with trigger and zoom
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_UTIL_OSCILLOSCOPE_H
#define MB_UTIL_OSCILLOSCOPE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbUtilOscilloscope : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-util-oscilloscope";
    static constexpr const char* PLUGIN_NAME    = "MB Oscilloscope";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float timeScale = 10f;  // range [1, 100]
    float triggerLevel = 0f;  // range [-1, 1]
    float zoom = 1f;  // range [0.1, 10]
    };

    MbUtilOscilloscope() = default;
    ~MbUtilOscilloscope() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.timeScale = std::clamp(params.timeScale, 1f, 100f);
        params.triggerLevel = std::clamp(params.triggerLevel, -1f, 1f);
        params.zoom = std::clamp(params.zoom, 0.1f, 10f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Oscilloscope
        return input;
    }
};

#endif // MB_UTIL_OSCILLOSCOPE_H

File: built-in plugins dsp/mb-util-oscilloscope.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbUtilOscilloscopePlugin: PluginDefinition = {
  id: "mb-util-oscilloscope",
  slug: "mb-util-oscilloscope",
  name: "MB Oscilloscope",
  category: "effect",
  type: "eq" as any,
  version: "1.0.0",
  description: "Real-time waveform display with trigger and zoom",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "timeScale",
      name: "Time Scale",
      type: "float",
      defaultValue: 10,
      minValue: 1,
      maxValue: 100,
      automatable: false,
    },
    {
      id: "triggerLevel",
      name: "Trigger Level",
      type: "float",
      defaultValue: 0,
      minValue: -1,
      maxValue: 1,
      automatable: false,
    },
    {
      id: "zoom",
      name: "Zoom",
      type: "float",
      defaultValue: 1,
      minValue: 0.1,
      maxValue: 10,
      automatable: false,
    },
  ],
  defaultPreset: { timeScale: 10, triggerLevel: 0, zoom: 1 },
};

export default MbUtilOscilloscopePlugin;
